### PR TITLE
More detailed unit info and small improvements

### PR DESCRIPTION
**Pull request by KionX**
UEF ACU

Cybran SACU

Presets

Partially improved describe presets and enhancements.
Impossible to fully correctly describe presets and enhancements. There is not enough data in the blueprint.

### CODE REVIEW

File: loc/US/strings_db.lua [hunk 1]
```diff
@@ -3686,14 +3688,15 @@ ability_massive="Massive"
 ability_notcap="Not Capturable"
 ability_omni="Omni Sensor"
 ability_personalshield="Personal Shield"
-ability_personalstealth="Personal Stealth"
+ability_personalstealth="Radar Stealth"
```
Comment (shalkya) [inline loc/US/strings_db.lua:3691]: "Personal Stealth" is a useful name
"Radar Stealth" isn't

File: loc/US/strings_db.lua [hunk 2]
```diff
@@ -3686,14 +3688,15 @@ ability_massive="Massive"
...
 ability_radar="Radar"
 ability_reclaim="Reclaims"
 ability_repairs="Repairs"
 ability_sacrifice="Sacrifice"
 ability_shielddome="Shield Dome"
 ability_sonar="Sonar"
-ability_stealthfield="Stealth Field"
+ability_stealth_sonar="Sonar Stealth"
+ability_stealthfield="Radar Stealth"
```
Comment (shalkya) [inline loc/US/strings_db.lua:3699]: "Stealth Field" is a useful name
"Sonar Stealth" and "Radar Stealth" aren't. I don't see the point of making a radar and a sonar stealth. For now only 1 stealth exist, and it work against sonar and radar.

File: loc/US/strings_db.lua [hunk 3]
```diff
@@ -3704,6 +3707,10 @@ ability_torpedo="Torpedoes"
 ability_torpedodef="Torpedo Defense"
 ability_transport="Transport"
 ability_upgradable="Upgradeable"
+ability_flying="Flying"
+ability_transportable="Transportable"
+ability_construction="Construction"
```
Comment (shalkya) [inline loc/US/strings_db.lua:3712]: it's the ability for a unit that can build stuff right ? then i would name it "build" not "construction"

File: loc/US/strings_db.lua [hunk 4]
```diff
@@ -6153,7 +6160,7 @@ ueb1102_help="Hydrocarbon Power Plant"
 ueb1102_name="HCPP - X1000"
 ueb1103_desc="Mass Extractor"
 ueb1103_help="Mass Extractor"
-ueb1103_name="Mass Pump 1"
```
Comment (shalkya) [inline loc/US/strings_db.lua:6156]: i wouldn't remove the number

File: loc/US/strings_db.lua [hunk 5]
```diff
@@ -6165,13 +6172,13 @@ ueb1201_help="Power Generator"
 ueb1201_name="EG - 200 Fusion Reactor"
 ueb1202_desc="Mass Extractor"
 ueb1202_help="Mass Extractor"
-ueb1202_name="Mass Pump 2"
```
Comment (shalkya) [inline loc/US/strings_db.lua:6168]: i wouldn't remove the number

File: loc/US/strings_db.lua [hunk 6]
```diff
@@ -6165,13 +6172,13 @@ ueb1201_help="Power Generator"
...
+ueb1202_name="Mass Pump"
 ueb1301_desc="Power Generator"
 ueb1301_help="Power Generator"
 ueb1301_name="EG 900 Fusion Reactor"
 ueb1302_desc="Mass Extractor"
 ueb1302_help="Mass Extractor"
-ueb1302_name="Mass Pump 3"
```
Comment (shalkya) [inline loc/US/strings_db.lua:6174]: i wouldn't remove the number

File: loc/US/strings_db.lua [hunk 7]
```diff
@@ -7,85 +7,85 @@
 --Aeon
 ual0301_Engineer_desc="Support Armoured Command Unit (Engineer Preset)"
 ual0301_Engineer_help="Enhanced during construction with Rapid Fabricator."
-ual0301_Engineer_name="SACU (Engineer Preset)"
```
Comment (shalkya) [inline loc/US/strings_db.lua:10]: i wouldn't remove those, but i would only show that when trying to build scu in the quantum gateway

File: loc/US/strings_db.lua [hunk 8]
```diff
@@ -6654,6 +6661,48 @@ uvd_0006="Storage"
...
+at_ASF="ASF"
+at_TMD="TMD"
+
+an_Normal="Normal"
+an_Overcharge="Overcharge"
+an_Deathnuke="Deathnuke"
+an_ExperimentalFootfall="Experimental Footfall"
+an_FireBeetleExplosion="FireBeetle Explosion"
+an_CzarBeam="Czar Beam"
+an_OtheTacticalBomb="Othe Tactical Bomb"
```
Comment (shalkya) [inline loc/US/strings_db.lua:6703]: othe ?

File: loc/US/strings_db.lua (hunk 1 above)
Comment (KionX) [reply loc/US/strings_db.lua:3691]: Now the radius is written for the fields.

File: loc/US/strings_db.lua (hunk 2 above)
Comment (KionX) [reply loc/US/strings_db.lua:3699]: The engine distinguishes between Radar and Sonar.
This is why it is written separately.
Unit abilities may change in the future.
Radar and Sonar may differ in mods.

File: loc/US/strings_db.lua (hunk 3 above)
Comment (KionX) [reply loc/US/strings_db.lua:3712]: This is referred to a lot as 'CONSTRUCTION'.
I just followed this tradition. If "build" looks better then I will do so.

File: loc/US/strings_db.lua (hunk 4 above)
Comment (KionX) [reply loc/US/strings_db.lua:6156]: "Mass Pump 1", "Mass Pump 2", "Mass Pump 3" what's the point of them?
The numbers are the same as the tech level. "Mass Pump" is a synonym for "Mass Extractor"

File: loc/US/strings_db.lua (hunk 7 above)
Comment (KionX) [reply loc/US/strings_db.lua:10]: name + desc is too long text.
name partially repeats desc.
With the removal of name, information is not lost.

File: loc/US/strings_db.lua (hunk 8 above)
Comment (KionX) [reply loc/US/strings_db.lua:6703]: These are the effects of armor as they are called in Lua.
"Othe Tactical Bomb" is name an weapon of Ahwassa.

**Comment by shalkya** [comment]
image
I think both stealth should be shown listed together (not separated by the sonar value), anyway i'm explaining later why i would prefer radar/sonar stealth to be removed.
image
Not true, it's sonar stealth & radar stealth. I also find it more informative to have stealth field here.
image
This is not only radar stealth, but also sonar stealth.
image
image
for these 2, same thing : radar stealth and sonar stealth. Also stealth field is more informative.
image
this one should have radar stealth and sonar stealth. Also personal stealth is more informative.

Overall i understand why you made radar stealth and sonar stealth and removed personal stealth and stealth field for the "more unit description" feature. But it leads to inaccurate information or lack of information outside of the "more unit description" feature (Also for the unit database). The sonar stealth and the radar stealth does indeed exist. But afaik only 1 unit in the game is using 2 different values for them (the T3 cybran sonar) and there is high chances that balance team is going to make these value the same, for practical reason. Because of that I would like you to change the work you did for stealth, and keep it as before.

image
i wouldn't show "T3 Support Armoured Command Unit (Combatant Preset)' (which is pretty long) but only "SACU (Combatant Preset)". I think you should use for sacu preset, the name of the preset (eg the thing you removed).

**Comment by shalkya** [comment]
image
image
image

I know what you tried to mean with "less is better". But it isn't very explicit. I would advise you to change to "% of damage taken"
and then you display "Normal : 100%" "Overcharge : 25%" "Deathnuke : 3.2%" "CZAR beam : 25%" "Othe Tactical Bomb : 10%"

Maybe it would be even better to do "% of resistance"
"Normal : 0 %" "Overcharge : 75%" "Deathnuke "96.8%" "CZAR beam : 75%" "Othe Tactical Bomb : 90%"

Btw you really need to change "Othe Tactical Bomb" to "Ahwassa Bomb". Nobody will understand what "Othe Tactical Bomb" is.

File: loc/US/strings_db.lua [hunk 9]
```diff
@@ -6654,6 +6661,48 @@ uvd_0006="Storage"
...
+uvd_basic="Basic"
+uvd_upgrades="Upgrades"
+
+at_Default="Default"
+at_Normal="Normal"
+at_Light="Light"
+at_Commander="Commander"
+at_Structure="Structure"
+at_Experimental="Experimental"
+at_FireBeetle="FireBeetle"
```
Comment (shalkya) [inline loc/US/strings_db.lua:6693]: I am removing the beetle armor type here #3103
I think you should also remove it here.

**Comment by shalkya** [comment]
image
image
it's radar stealth and sonar stealth
i still feels displaying "personal radar stealth" or "personal sonar stealth" is a bit better. In the same way it works for shield.
image
image
it's radar stealth field and sonar stealth field
image
the sacu (preset or not) have wrong information. 
1st : They all display the same information
2nd : Some have ability displayed, that they shouldn't have. Here the "cloacking" is only for cloack preset, and shouldn't appear for others sacu.
3rd : Some ability aren't shown. For example the Anti-air preset, should have the "anti-air" ability. It doesn't have it right now.
Edit : dps aren't updated either. It's mostly wrong for all preset. This need to be fixed, or hide information (if you can't fix) and show only the explanation text. We shouldn't display wrong information.
image
the "anti-shield" dps isn't shown, while it's the most important information.

**Comment by shalkya** [comment]
image
the death nuke damage is wrong. Should be around 2.5k and range is wrong too.
it lacks other upgrades explanation. Could at least display the gun, since you display the other offensive upgrades.
the "Nuclear warhead" information is wrong. Wrong damage and splash.

Overall the ACU information display useless info (tele damage for example) while ignoring the upgrade survivability upgrades (nano, shields etc etc)

**Comment by KionX** [comment]
```They all display the same information```
Because the blueprint is the same.
```Here the "cloacking" is only for cloack preset```
Because the blueprint is the same and cloak exists.
```Anti-air preset, should have the "anti-air" ability```
I did not make abilities for weapons. This ability can be added to the blueprint.
```dps aren't updated either```
It is not possible to apply the enhancements effects before creating a unit.
```the "anti-shield" dps isn't shown```
```the "Nuclear warhead" information is wrong```
This is a specific effects.

If all special cases are taken into account, the code will turn into fucked shit.

**Comment by KionX** [comment]
Impossible to correctly describe presets. There is not enough data in the blueprint.
I suggest deleting the simple list of abilities.

**Comment by shalkya** [comment]
image
no stealth information on harms

can't build ML and megalith anymore.

errors : 
```
WARNING: Error running HandleEvent script in CScriptObject at 1a3dc140: ...rs\pierre-yves\documents\github\fa\lua\maui\text.lua(172): bad argument #1 to `gfind' (string expected, got nil)
         stack traceback:
         	[C]: in function `gfind'
         	...rs\pierre-yves\documents\github\fa\lua\maui\text.lua(172): in function <...rs\pierre-yves\documents\github\fa\lua\maui\text.lua:151>
         	...s\documents\github\fa\lua\ui\game\unitviewdetail.lua(427): in function `WrapAndPlaceText'
         	...s\documents\github\fa\lua\ui\game\unitviewdetail.lua(687): in function `Show'
         	...ves\documents\github\fa\lua\ui\game\construction.lua(1159): in function `OnRolloverEvent'
         	...\pierre-yves\documents\github\fa\lua\maui\button.lua(62): in function <...\pierre-yves\documents\github\fa\lua\maui\button.lua:50>
WARNING: Error running HandleEvent script in CScriptObject at 1a3dc140: ...rs\pierre-yves\documents\github\fa\lua\maui\text.lua(172): bad argument #1 to `gfind' (string expected, got nil)
```
When hovering over Harms icon, when selecting it
```
WARNING: Error running HandleEvent script in CScriptObject at 225fe880: ...s\documents\github\fa\lua\ui\game\unitviewdetail.lua(411): bad argument #4 to `format' (number expected, got no value)
         stack traceback:
         	[C]: in function `format'
         	...s\documents\github\fa\lua\ui\game\unitviewdetail.lua(411): in function `?'
         	...s\documents\github\fa\lua\ui\game\unitviewdetail.lua(449): in function `WrapAndPlaceText'
         	...s\documents\github\fa\lua\ui\game\unitviewdetail.lua(687): in function `Show'
         	...ves\documents\github\fa\lua\ui\game\construction.lua(1157): in function `OnRolloverEvent'
         	...\pierre-yves\documents\github\fa\lua\maui\button.lua(62): in function <...\pierre-yves\documents\github\fa\lua\maui\button.lua:50>
```
When hovering over ML icon when selecting it

Can't get the info about Harms, ML and mega when hovering over their icon

**Comment by KionX** [comment]
```no stealth information on harms```
Xrb2308 doesn't have stealth. Then it is replaced with xrb2309.

**Comment by shalkya** [comment]
alright so pretty much all good.
Last thing i would like is the death damage / crash damage of experimental (land and air). I think it's important and it's for several units.

**Comment by KionX** [comment]
Side effect - showing teleport ACU as weapon.
Can't hide it. The reason is the DummyWeapon flag.
Perhaps some more DummyWeapon will be displayed.

**Comment by shalkya** [comment]
yeah well it's not that bad.
Zthuue shows 15 dps and lobo shows 12 dps. I think it's due to them splitting in several bomb (a volley of multiple projectile). I think it's pretty bad if we display wrong dps, can you do something about that ?

**Comment by KionX** [comment]
I did not find a way to improve calculate the DPS.

**Comment by KionX** [comment]
Reasons:
Unknown how many projectiles are in the salvo.
Unknown how many projectiles are in the queue.
Unknown how many parts the projectiles are divided into.

Maybe this can be corrected through the blueprint.

**Comment by shalkya** [comment]
so rapidly i looked into the unit databases : 
for the normal unit database : https://github.com/FAForever/UnitDB
"T1 uef arty : DPS 12" WRONG
"T1 sera arty : DPS 16" WRONG
"T2 uef MML : DPS 60" TRUE
"salvation : DPS 71" WRONG
"novax : DPS 243" TRUE
"T2 sera PD : DPS 151" TRUE

for spooky unit database : https://github.com/spooky/unitdb
"T1 uef arty : DPS 60.24" TRUE
"T1 sera arty : DPS 77.59" TRUE
"T2 uef MML : DPS 60" TRUE
"salvation : DPS 70.97" WRONG
"novax : DPS 243" TRUE
"T2 sera PD : DPS 151.25" TRUE

You should look into their code to see how they calculate dps so you can improve the calculation in game. It seems that spooky can calculate T1 arty dps while normal database can't. Both can't calculate salvation dps.

**Comment by shalkya** [comment]
now other unit have wrong dps :
T2 uef torpedo launcher : 600 dps instead of expected 150 dps
UEF sam : 2000 dps instead of expected 333 dps
T1 cybran static AA : 140 dps instead of expected 70 dps
T2 cybran static flak : 300 dps instead of expected 150 dps
cybran sam : 1333 dps instead of expected 333 dps
harms : 2250 dps instead of expected 375 dps
T1 sera static AA : 133 dps instead of expected 67 dps
T2 sera torpedo launcher : 450 dps instead of expected 150 dps
T2 sera static flak : 285 instead of expected 140 dps
sera sam : 666 dps instead of expected 333 dps
T2 aeon torpedo launcher : 600 dps instead of expected 150 dps
aeon sam : 666 dps instead of expected 333 dps

**Comment by shalkya** [comment]
i guess you can remove dps for kamikaze (doesn't make sense), after that it should be good to merge.

**Comment by shalkya** [comment]
https://github.com/FAForever/fa/pull/3060/commits/fca09508a600c2c6ff1daf4bf26a27607b19cd3b did you only add the projectile files, or did you edit them at the same time ?

**Comment by KionX** [comment]
Edited a bit. I forgot to edit it as a separate commit.
~4 lines each to support new parameters.

**Comment by shalkya** [comment]
I understand you remade all the commits for sanity and to fix the commit mixing addition of files and modification of the same files, and i thank you for it.
But please, be careful !
https://github.com/FAForever/fa/pull/3060/commits/92e3faa39f14b023ff1c2260f97b5bd9097ff10b#diff-99a5c94d2d8bb8ecc67e191acf0f2ed2L699
You are reverting the change made to ythota death weapon !
I will keep looking in the files to spot any other revert, but can you also do the same, and prevent revert to sneak into your pr in the future ? 
I am not sure what caused it, but most likely you have some habits using git they may have caused it. For example, it's most likely unrelated i guess, but the pr should not need to be closed and reopened. You already told me it was automatic, but this isn't supposed to happen, even if you intend to force push.

**Comment by KionX** [comment]
I checked the PR again and corrected all unnecessary changes.

**Comment by shalkya** [comment]
thanks i'll look into it

**Pull request merged**